Riot/Modules/Room/VoiceMessages/VoiceMessageAudioRecorder.swift

```diff
@@ -0,0 +1,141 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import AVFoundation
+
+protocol VoiceMessageAudioRecorderDelegate: AnyObject {
+    func audioRecorderDidStartRecording(_ audioRecorder: VoiceMessageAudioRecorder)
+    func audioRecorderDidFinishRecording(_ audioRecorder: VoiceMessageAudioRecorder)
+    func audioRecorder(_ audioRecorder: VoiceMessageAudioRecorder, didFailWithError: Error)
+}
+
+enum VoiceMessageAudioRecorderError: Error {
+    case genericError
+}
+
+class VoiceMessageAudioRecorder: NSObject, AVAudioRecorderDelegate {
+    
+    private enum Constants {
+        static let silenceThreshold: Float = -50.0
+    }
+    
+    private var audioRecorder: AVAudioRecorder?
+    private let delegateContainer = DelegateContainer()
+    
+    var url: URL? {
+        return audioRecorder?.url
+    }
+    
+    var currentTime: TimeInterval {
+        return audioRecorder?.currentTime ?? 0
+    }
+    
+    var isRecording: Bool {
+        return audioRecorder?.isRecording ?? false
+    }
+    
+    func recordWithOuputURL(_ url: URL) {
+        
+        let settings = [AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
```

Comment:
Probably not an issue, but as an audio nerd it bothers me that this records to a lossy format rather than lossless when it's going to be converted to another lossy format immediately before sending. 🤓

Reply:
Talked about this privatelly and decided to try to only switch the container format when converting, using opus here.